ipykernel/kernelbase.py

```diff
@@ -214,7 +211,7 @@ def process_control(self, msg):
             self.log.error("UNKNOWN CONTROL MESSAGE TYPE: %r", msg_type)
         else:
             try:
-                yield gen.maybe_future(handler(self.control_stream, idents, msg))
+                await handler(self.control_stream, idents, msg)
```

Comment:
This is a backward-incompatible change to *force* all handlers to be async. We could keep the maybe_future functionality with [inspect.isawaitable](https://docs.python.org/3/library/inspect.html#inspect.isawaitable):

```suggestion
                result = handler(self.control_stream, idents, msg)
                if inspect.isawaitable(result):
                    await result
```

Reply:
I like this solution.

Quick thought: should we try to avoid a situation in which some handlers are async and others not?

Reply:
> should we try to avoid a situation in which some handlers are async and others not?

I don't think it's important, as long as all handlers are *allowed* to be async.

Reply:
I added your suggestion for both shell and control handlers, raison a deprecation warning when the handlers is not awaitable.